src/utils/code-import.ts

```diff
@@ -0,0 +1,145 @@
+import fs from "node:fs";
+import path from "node:path";
+import { EOL } from "node:os";
+import { visit } from "unist-util-visit";
+import stripIndent from "strip-indent";
+import type { Root, Code, Parent } from "mdast";
+import type { VFile } from "vfile";
+
+interface CodeImportOptions {
+  async?: boolean;
+  preserveTrailingNewline?: boolean;
+  removeRedundantIndentations?: boolean;
+  rootDir?: string;
+  allowImportingFromOutside?: boolean;
+}
+
+function extractLines(
+  content: string,
+  fromLine: number | undefined,
+  hasDash: boolean,
+  toLine: number | undefined,
+  preserveTrailingNewline: boolean = false,
+) {
+  const lines = content.split(EOL);
+  const start = fromLine || 1;
+  let end;
+  if (!hasDash) {
+    end = start;
+  } else if (toLine) {
+    end = toLine;
+  } else if (lines[lines.length - 1] === "" && !preserveTrailingNewline) {
+    end = lines.length - 1;
+  } else {
+    end = lines.length;
+  }
+  return lines.slice(start - 1, end).join("\n");
+}
+
+function codeImport(options: CodeImportOptions = {}) {
```

Comment:
This is nitpicky, but `importCode` may be better (function should be verbs, see https://stackoverflow.com/questions/1991324/function-naming-conventions for some good discussion).